### ISSUE TITLE
[Cuda Stream/Event] Python API acces to cudaStreamWaitEvent

### ISSUE THREAD

**Issue by gauenk** [pytorch/pytorch#67866]
## 🚀 Feature

I would like access to the "cudaStreamWaitEvent" through the Python API, maybe specifically through the Stream or Event API.

## Motivation

I would like to synchronize between CUDA streams. 

## Pitch

Synchronizing between streams is a common access pattern. Say we want to assign a different stream to different sections of an image, and we do this repeatedly.

## Alternatives

There seem to be two locations where the cudaStreamWaitEvent would be accessible: the CUDAEvent::block function and CUDAStream::synchronize_with in the pytorch/aten/src/ATen/cuda directory.

Line 180 of pytorch/c10/cuda/CUDAStream.h says:
```
// Deleted for now; use CUDAEvent::block instead                                         
// void synchronize_with(const CUDAEvent& event) const;  
```
However, the CUDAEvent's block function is not currently accessible through the Python API. 

## Additional context

I want to use CUDA streaming as done in FAISS. An example of the streaming is shown [in this file](https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/impl/Distance.cu), lines 236, 237, and 400. The synchronization pattern is shown [in this file](https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/utils/DeviceUtils.h) in function named streamWaitBase on line 157.

Thanks for considering!

cc @brianjo @mruberry @ngimel

**Comment by ngimel** [comment]
This functionality is available through cuda.Event.wait https://pytorch.org/docs/stable/generated/torch.cuda.Event.html?highlight=cuda%20event#torch.cuda.Event which calls event.block underneath.

**Comment by gauenk** [comment]
Thank you for your comment. I just checked and I confirm your comment. 

Should the call to cudaStreamWaitEvent be made explicit in the documentation, like how [Event.synchronize](https://pytorch.org/docs/stable/generated/torch.cuda.Event.html?highlight=cuda%20event#torch.cuda.Event) is wrapping cudaEventSynchronize? I just got lost in the nest of function calls (./torch/csrc/cuda/Event.cpp line 123 and ./aten/src/ATen/cuda/CUDAEvent.h line 128). Maybe I am silly, but maybe this is not clear to other users as well.
